### PR TITLE
Merge image models

### PR DESCRIPTION
**Pull request by symroe**
This is a little different than I assumed it would be:

1. The `Image` model is actually a 3rd party app, so merging it like the other models would require adding it to the project.
2. The only model that's using `Image` is `popolo.Person` via a GenericRelation. We're generally moving away from them anyway.
3. Because of this I decided to add a `PersonImage` model. 
4. To save adding this to the popolo or candidates app and then moving it later, I created a new app `people`. I plan to move `Person` there at some point in the future, but that's out of the scope of this PR.
5. Removing the generic Image model means removing images from `Organization` too. This means we can't serve party emblems for the `v0.9` API any more (well, not without a lot of effort). I think this is reasonable, given that we're not a canonical source for these images anyway. This will mean updating the WCIVF importer to use the `next` Party endpoint.
6. The 3rd party package added some nice admin includes that I've not duplicated here. I've never used the admin to work on images (it's sort of the point of the site that we can admin it in the frontend), so we can add these back in if we need them

### CODE REVIEW

**Comment by symroe** [comment]
Oh, one thing I missed from the descriptions is that we can't remove the `mysociety-django-images` dependency because some migrations rely on it existing.

These migrations will be squashed/deleted in the future, but until then we need to keep it installed.

**Comment by chris48s** [comment]
As written, if we run the migrations in this PR it will delete all our stuff and then break:

```
Applying candidates.0050_delete_imageextra... OK
Applying people.0001_initial... OK
Applying popolo.0022_popolate_person_image_from_image...Traceback (most recent call last):

<snip>

django.core.exceptions.FieldError: Invalid field name(s) given in select_related: 'extra'. Choices are: content_type
```

This is probably not the intended behaviour :) I think you need to change the migration dependencies so that `0022_popolate_person_image_from_image` runs _after_ `0050_delete_imageextra` not before it. I suspect this didn't happen for you because you ran the migrations as you were writing them not as a single block.

---

I suspect this block of 3 things is all related:

* `/moderation/photo/review` doesn't show any thumbnail images for me.

* If I approve an image upload, a records go into the `PersonImage` model, but it never appears on the candidate page. In the template block

    ```django
    {% if person.primary_image %}
      {% thumbnail person.primary_image "x80" as im %}
        <img class="person-avatar" src="{{ im.url }}"/>
      {% endthumbnail %}
    {% else %}
    ```

    `person.primary_image` is truthy, but no ` This will mean updating the WCIVF importer to use the `next` Party endpoint.

I don't see that as a bad thing. While we're talking API, I can see why you want to map `api/v0.9/images` to the `PersonImage` model, but do you think it would be better to mount the endpoint at `api/next/person_images` going forwards to give ourselves options..

---

This is something we can look at for future instead of right now (lets not scope creep in this PR), but it seems like we've got a bunch of code sprinkled around the place working out MD5 sums: https://github.com/DemocracyClub/yournextrepresentative/search?q=md5&unscoped_q=md5 Instead of requiring the calling object to pass an `md5sum` and then just saving it, do you think it there is any mileage in looking at consolidating some of that logic for calculating the hash and checking files against the checksum into the image class?

---

As you can see from the CI build there are some test failure to look at. I'll leave a few other comments inline on some more minor stuff

File: ynr/apps/people/tests.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
Comment (chris48s) [inline ynr/apps/people/tests.py]: For reasons noted in the previous PR, lets delete this

File: ynr/apps/candidates/migrations/0050_delete_imageextra.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    dependencies = [("candidates", "0049_cleanup_post_extra")]
+
+    operations = [
+        migrations.RemoveField(model_name="imageextra", name="base"),
+        migrations.RemoveField(model_name="imageextra", name="uploading_user"),
+        migrations.RemoveField(model_name="organizationextra", name="base"),
+        migrations.AlterField(
+            model_name="extrafield",
+            name="type",
```
Comment (chris48s) [inline ynr/apps/candidates/migrations/0050_delete_imageextra.py]: I'm confused about why this changes in this migration. Can you clarify?

File: ynr/apps/popolo/migrations/0022_popolate_person_image_from_image.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
```
Comment (chris48s) [inline ynr/apps/popolo/migrations/0022_popolate_person_image_from_image.py]: Should the name of this file be `populate_person_image_from_image`, or was the pun intentional? :D

File: ynr/apps/candidates/tests/test_constituency_view.py [hunk 4]
```diff
@@ -96,7 +96,7 @@ def test_any_constituency_page_without_login(self):
 
     def test_any_constituency_page(self):
         # Just a smoke test for the moment:
-        with self.assertNumQueries(46):
+        with self.assertNumQueries(48):
```
Comment (chris48s) [inline ynr/apps/candidates/tests/test_constituency_view.py]: I'm confused about why removing the one-to-one and generic relations here means we're making _more_ queries. Did you look into this at all?

**Comment by chris48s** [comment]
Oh yeah, one more thing. In order to approve image uploads locally, I had to set

`EMAIL_BACKEND = 'django.core.mail.backends.dummy.EmailBackend'`

Shall we shove this in `local.example.py`. I reckon that's a sensible default for a dev install

File: ynr/apps/candidates/migrations/0050_delete_imageextra.py (hunk 2 above)
Comment (symroe) [reply ynr/apps/candidates/migrations/0050_delete_imageextra.py]: This was a mistake on my part because there are still python 3 related unicode migrations left to apply, and I didn't remove it.

I think I'll just do those migrations in a single PR after this, to save me having to pick them out all the time 🙄

File: ynr/apps/popolo/migrations/0022_popolate_person_image_from_image.py (hunk 3 above)
Comment (symroe) [reply ynr/apps/popolo/migrations/0022_popolate_person_image_from_image.py]: Ha! It's a nice pun, but I changed it

**Comment by symroe** [comment]
* > relation "thumbnail_kvstore" does not exist
   
   I suspect this isn't a new problem with this PR. Did it work before? After some searching, it looks like this is the problem: https://stackoverflow.com/questions/35136411/table-thumbnail-kvstore-doesnt-exist. Maybe rebase over #644 or install 12.4 on this branch and migrate? I guess I'm not seeing it because I've not started with a blank DB since an earlier version of sorl?

   Related, `{% load thumbnail %}` is from `sorl.thumbnail`. I've been wanting to DRY up that code for ages, so I gave myself permission to widen the scope slightly in this PR in 83a44f6

* On  the `EMAIL_BACKEND`: I've set this to `django.core.mail.backends.console.EmailBackend` as I think that's generally more useful for testing (outputs to the `runserver` console, rather than just `/dev/null`)

* On `api/next/person_images` – this implies it's useful for someone to iterate over all person images without wanting to iterate over all people. I'm not sure this is very useful, but if you can suggest a use case then I'm happy to add it. I think the previous endpoint was more useful because `Image` was a generic concept

**Comment by chris48s** [comment]
You're right about the thumbnails being an issue with sorl. Upgrading to sorl 12.4.1 and running migrations sorted that :+1:



Migrations seem to now be applying in a sensible order.



> On the `EMAIL_BACKEND`: I've set this to `django.core.mail.backends.console.EmailBackend`

Yep good call.



> On `api/next/person_images` – this implies it's useful for someone to iterate over all person images without wanting to iterate over all people. I'm not sure this is very useful, but if you can suggest a use case then I'm happy to add it. I think the previous endpoint was more useful because Image was a generic concept

Well we don't need to add anything - its already there, its just that it is called `api/next/images`:

screenshot at 2018-09-24 10-20-24

My original thought was that if we're going to keep it, should we _rename_ it.. but you do raise a good point: I think you're right that it would be better to completely _remove_ that endpoint from `/next` and just serialize `PersonImages` as an `images` array key inline in a person object :+1: That makes more sense.

I'm a bit confused though, because locally I now have `thumbnail` and `image` keys in the API outputs, but no `images` array. However looking at the live site there is an images array key in the API output. Do you have an `images` array in the person object in your local API outputs? From the API code it looks like it should be there..

**Comment by chris48s** [comment]
clarification:

* I have an `images[]` key locally on the `/v0.9` endpoint but not `/next` locally
* Its there on both `/v0.9` and `/next` in production

**Comment by chris48s** [comment]
>  Fix party importing from live site
> This change is a little hard to test, as the legacy_slug is required to
> make the models, but we're not exposing it in the API yet.

I tried to rebuild to test the migrations and hit this exact same issue :) Was going to raise it but I expect this will fix it (like you say, hard to test!). I'll try rebuilding again once we deploy this to check it.

File: ynr/apps/elections/uk/templates/candidates/person-view.html [hunk 5]
```diff
@@ -49,14 +49,11 @@
...
-    {% else %}
-      {% if user.is_authenticated %}
-        <a class="upload-photo" href="{% url 'photo-upload' person.id %}">
-          {% trans 'Upload photo' %}
-        </a>
-      {% endif %}
+    <img class="person-avatar" src="{{ person.get_display_image_url }}"/>
+    {% if not person.primary_image and user.is_authenticated %}
+      <a class="upload-photo" href="{% url 'photo-upload' person.id %}">
+      w  {% trans 'Upload photo' %}
```
Comment (chris48s) [inline ynr/apps/elections/uk/templates/candidates/person-view.html]: What is the significance of the 'w' character here?

File: ynr/apps/popolo/migrations/0022_populate_person_image_from_image.py [hunk 6]
```diff
@@ -38,7 +38,7 @@ def popolate_person_image_model(apps, schema_editor):
             is_primary=image.is_primary,
         )
     # Make sure we've copied all the images
-    assert PersonImage.objects.count() == image_qs.count()
+    assert PersonImage.objects.count() >= image_qs.count()
```
Comment (chris48s) [inline ynr/apps/popolo/migrations/0022_populate_person_image_from_image.py]: How do we end up with more records than we started with here?

File: ynr/apps/elections/uk/templates/candidates/person-view.html [hunk 7]
```diff
@@ -62,21 +48,12 @@
...
-      {% if user.is_authenticated %}
-        <a class="upload-photo" href="{% url 'photo-upload' person.id %}">
-          {% trans 'Upload photo' %}
-        </a>
-      {% endif %}
+
+    <img class="person-avatar" src="{{ person.get_display_image_url }}"/>
+    {% if not person.primary_image and user.is_authenticated %}
+      <a class="upload-photo" href="{% url 'photo-upload' person.id %}">
+      w  {% trans 'Upload photo' %}
```
Comment (chris48s) [inline ynr/apps/elections/uk/templates/candidates/person-view.html]: What is the significance of the 'w' character here?

File: ynr/apps/popolo/migrations/0022_populate_person_image_from_image.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            copyright=image.extra.copyright,
+            uploading_user=image.extra.uploading_user,
+            user_notes=image.extra.user_notes,
+            md5sum=image.extra.md5sum,
+            user_copyright=image.extra.user_copyright,
+            notes=image.extra.notes,
+            is_primary=image.is_primary,
+        )
+    # Make sure we've copied all the images
+    assert PersonImage.objects.count() >= image_qs.count()
```
Comment (chris48s) [inline ynr/apps/popolo/migrations/0022_populate_person_image_from_image.py]: How do we end up with more records than we started with here?

**Comment by chris48s** [comment]
Applying commit 0949cb8 causes calls to the `/persons` API endpoint to fail with `Could not resolve URL for hyperlinked relationship using view name "personimage-detail"` because we're using a hyperlinked serializer, but the thing we're hyperlinking to has been deleted.

---

In general, it seems like we are lacking tests for the API:

* One of the changes in this PR caused us to unexpecedly lose a key from our API outputs (now fixed). All the tests passed.
* Another change ( 0949cb8 ) causes every request to the `/persons` endpoint to throw an exception. All the tests still pass.
* Not _directly_ related to this PR as it happened some time ago, but if we had any test for the images endpoint, at the point we upgraded sorl to the buggy release with no migrations, that should have also caused a failure in CI.

I don't think "add an entire integration test suite for the API" is good scope creep for this PR, but even a test suite which just calls each API endpoint and checks the status is `200 OK` would have picked up 2 of those 3 issues. Given how heavily we're leaning on the test suite at the moment to support refactoring, it feels like this is an area where we are missing some coverage and could really benefit from adding more tests.

---

Bikeshed: There's a spellin n grammer fixup needed in 
8b8a51f : "jsoin.loads if the obkject exists" --> "json.loads if the object exists"

File: ynr/apps/elections/uk/templates/candidates/person-view.html (hunk 7 above)
Comment (symroe) [reply ynr/apps/elections/uk/templates/candidates/person-view.html]: Whoops!

**Comment by symroe** [comment]
Smoke tests added, and it caught a couple of things that I've also fixed. Obviously more tests are important, but I think this is an improvement for now. Thanks!

**Comment by chris48s** [comment]
`/api/next/persons/` now works for me, but `/api/v0.9/persons/` is still throwing `Could not resolve URL for hyperlinked relationship using view name "personimage-detail"` because we're still declaring `class ImageSerializer(serializers.HyperlinkedModelSerializer)` in `ynr/apps/api/v09/serializers.py` (should it just be importing the serializer from the people app?). This also implies that the test added in f727082 isn't working as expected or is not being run at all. The build should now be failing as a result of this issue.

File: ynr/apps/api/tests/test_api_smoke_tests.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    def test_each_endpoint(self):
+        endpoints_to_ignore = ["candidates_for_postcode"]
+        for version in ("v0.9", "next"):
+            response = self.app.get("/api/{}/".format(version))
+            for name, url in response.json.items():
+                # This slightly odd dance is needed because
+                # pytest doesn't actually support subtests yet
+                # (see https://github.com/pytest-dev/pytest/issues/1367)
+                # Use try/except/failureException to log the endpoint that
+                # failed. Unlike proper subtests, this will tail the test,
```
Comment (chris48s) [inline ynr/apps/api/tests/test_api_smoke_tests.py]: "this will tail the test" --> "this will fail the test"

**Comment by chris48s** [comment]
One final v. minor. The commit message in 8f46855  still says "**jsoin**.loads"

**Comment by symroe** [comment]
Thanks, the `v0.9` person endpoint wasn't failing because I'd not made a `PersonImage` for the smoke tests. I've duplicated the test in the `next` endpoint that checks for the image.

More generally, I think making sure we have a good set of test fixtures is a good idea, and something I'll look at in future. There are loads of tests that duplicate setting up realistic memberships and other data.
